Sources/StreamChat/AudioPlayer/AudioPlaying.swift

```diff
@@ -0,0 +1,375 @@
+    /// Provides  way to get an instance of a player
+    static func build() -> AudioPlaying
+
+    /// Requests the playbackContext for the given URL. If the player's current item has as URL that
+    /// matches the provided one, it should return a context, otherwise it will return
+    /// ``AudioPlaybackContext.notLoaded``
+    /// - Parameter url: The URL (provided by the asset) that is used to stream/download the content to play
+    /// - Returns: if the provided URL matches the current item's URL, otherwise it will return the context
+    /// for the current playback. Otherwise, the context will be ``.notLoaded``.
+    func playbackContext(for url: URL) -> AudioPlaybackContext
+
+    /// Instructs the player to load the asset from the provided URL and prepare it for streaming. If the
+    /// player's current item has a URL that matches the provided one, then we will try to play or restart
+    /// the playback while updating the new delegate.
+    /// - Parameters:
+    ///   - url: The URL where the asset will be streamed from. If nil then the player will simply clear up the current playback queue.
+    ///   - delegate: The delegate that will be informed for changes on the asset's playback.
+    func loadAsset(from url: URL?, andConnectDelegate delegate: AudioPlayingDelegate)
+
+    /// Begin the loaded asset's playback. If no asset has been loaded, the action has no effect
+    func play()
+
+    /// Pauses the loaded asset's playback. If non has been loaded or the playback hasn't started yet
+    /// the action has no effect.
+    func pause()
+
+    /// Stop the loaded asset's playback. If non has been loaded or the playback hasn't started yet
+    /// the action has no effect.
+    func stop()
+
+    /// Updates the loaded asset's playback rate to the provided one.
+    /// - Parameter newRate: The new rate which we want the player to use for playback
+    func updateRate(_ newRate: AudioPlaybackRate)
+
+    /// Performs a seek in the loaded asset's timeline at the provided time.
+    /// - Parameter time: The time to seek at
+    func seek(to time: TimeInterval)
+}
+
+/// An implementation of ``AudioPlaying`` that can be used to stream audio files from a URL
+open class StreamRemoteAudioPlayer: AudioPlaying {
+    // MARK: - Properties
+
+    /// The queue used to ensure thread-safe access to the context property
+    private lazy var contextAccessQueue: DispatchQueue = .init(
+        label: "com.getstream.audio.player.context", qos: .userInteractive
+    )
+    /// Describes the player's current playback state. The access to this property is **not** thread-safe
+    private var _context: AudioPlaybackContext = .notLoaded
+    /// Describes the player's current playback state. The access to this property is thread-safe
+    private(set) var context: AudioPlaybackContext {
+        get { contextAccessQueue.sync { _context } }
+        set { contextAccessQueue.sync { _context = newValue } }
```

Comment:
You can use `async` with a barrier flag here

Reply:
> The blocking of unwanted threads is a potential issue in the concurrent queue with the barrier task, right?
I mean that using `sync` for writing it is. Writing usually is a fire-and-forget operation, there's no need to wait for this to be completed. 

> The audio player is one and tied to the MessageListVC and is being reused for every message and the events are only coming from the player and not outside. I don't see the need for a concurrent queue here.
Using a concurrent queue could be a premature optimization indeed, but if we stick to the Serial Queue, I would still use `async` when changing the context 🤔

Reply:
I see your point, and I can change the writing to use `async` but still on the Serial Queue.